torchx/schedulers/kubernetes_mcad_scheduler.py

```diff
@@ -0,0 +1,1121 @@
+            start_index = role.args[1].find(rdzv)
+
+            # find first instance of : after rdzv endpoint
+            port_index = role.args[1].find(separator, start_index) + 1
+            end_index = role.args[1].find(space, port_index)
+            port = role.args[1][port_index:end_index]
+
+    return port
+
+
+def app_to_resource(
+    app: AppDef,
+    namespace: str,
+    service_account: Optional[str],
+    image_secret: Optional[str],
+    priority: Optional[int] = None,
+) -> Dict[str, Any]:
+    """
+    app_to_resource creates a AppWrapper/MCAD Kubernetes resource definition from
+    the provided AppDef. The resource definition can be used to launch the
+    app on Kubernetes.
+
+    """
+
+    genericitems = []
+
+    unique_app_id = cleanup_str(make_unique(app.name))
+    for role_idx, role in enumerate(app.roles):
+        for replica_id in range(role.num_replicas):
+            values = macros.Values(
+                img_root="",
+                app_id=unique_app_id,
+                replica_id=str(replica_id),
+                rank0_env=f"MCAD_{cleanup_str(app.roles[0].name)}_0_HOSTS".upper().replace(
+                    "-", ""
+                ),
+            )
+
+            if role_idx == 0 and replica_id == 0:
+                values.rank0_env = "TORCHX_RANK0_HOST"
+            name = cleanup_str(f"{unique_app_id}-{replica_id}")
+            replica_role = values.apply(role)
+            if role_idx == 0 and replica_id == 0:
+                replica_role.env["TORCHX_RANK0_HOST"] = "localhost"
+
+            pod = role_to_pod(
+                name,
+                unique_app_id,
+                namespace,
+                replica_role,
+                service_account,
+                image_secret,
+            )
+            pod.metadata.labels.update(pod_labels(app, role_idx, role, replica_id))
+
+            genericitem: Dict[str, Any] = {
+                "replicas": 1,
+                "generictemplate": pod,
+            }
+            # TODO: retry support here
+            if role.max_retries > 0:
+                genericitem["maxRetry"] = role.max_retries
+                genericitem["policies"] = RETRY_POLICIES[role.retry_policy]
+                msg = f"""
+Role {role.name} configured with restarts: {role.max_retries}.
+                """
+                warnings.warn(msg)
+
+            genericitems.append(genericitem)
+
+    """
+    Create Service:
+    The selector will have the key 'appwrapper.mcad.ibm.com', and the value will be 
+    the appwrapper name
+    """
+
+    service_port = get_port_for_service(app)
+
+    if not (0 < int(service_port) <= 65535):
+        msg = """Warning: rdzv_endpoint set to invalid port number. Value must be between 1-65535, with torchx default = 29500. Try updating --rdzv_port argument to dist.ddp. Setting port to default = 29500"""
+        service_port = "29500"
+        warnings.warn(msg)
+
+    svc_obj = mcad_svc(unique_app_id, namespace, service_port)
```

Comment:
what's this service used for? for pytorch rendezvous it's using the hostnames and not this right?

Reply:
This service is used for distributed jobs to enable the PyTorch rendezvous. Since MCAD treats the jobs holistically, it needs to know the Kubernetes service information and pod information and treats them as the same "AppWrapped" job. For example, a distributed job that will run as two pods will be represented in a single "App wrapper" as a pod with 2 replicas and 1 service. This includes needing the port that will be used for the PyTorch rendezvous. The Kubernetes Volcano scheduler automatically creates and destroys a service for distributed PyTorch jobs as well. However, Volcano manages this differently and so TorchX does not need to be aware of the Kubernetes service.

Reply:
is the service set as the hostname on the workers? I'm just not quite understanding why a service is needed here if the pods can directly talk to each other.  The service is for load balancing/discovery but we're using the pod host names so if I understanding correctly shouldn't be needed

Reply:
This section just generates the yaml specs for a Kubernetes service. TorchX-Volcano generates a similar Kubernetes Service as well, but because Volcano handles this automation the rest of TorchX does not even need to be aware of the Kubernetes service. MCAD does not have this type of automation in place, so it had to be moved to the TorchX scheduler level and included in the generate yaml file that is evaluated by MCAD. If one were to simply comment out this mcad service and try to run a distributed job, the main worker stalls at waiting to rendezvous with all workers and the replica workers will reach a time out.